File: apps/api/app/schemas/intake.py
from enum import Enum
from typing import Annotated

from pydantic import BaseModel, Field, HttpUrl, field_validator


class YearOfStudy(str, Enum):
    FIRST = "1st"
    SECOND = "2nd"
    THIRD = "3rd"
    FOURTH = "4th"
    DROPOUT = "Dropout"
    GAP_YEAR = "Gap year"


class TechStream(str, Enum):
    """Degree branch / stream (engineering, commerce, business, etc.)."""

    CS = "CS"
    IT = "IT"
    ECE = "ECE"
    EE = "EE"
    MECHANICAL = "Mechanical"
    CIVIL = "Civil"
    COMMERCE = "Commerce"
    BBA = "BBA"


class CollegeTier(str, Enum):
    TIER_1 = "Tier 1 (IIT/NIT/IIIT)"
    TIER_2 = "Tier 2"
    TIER_3 = "Tier 3"


class CgpaBand(str, Enum):
    BELOW_6_5 = "<6.5"
    BAND_6_5_7_5 = "6.5–7.5"
    BAND_7_5_8_5 = "7.5–8.5"
    ABOVE_8_5 = "8.5+"


class RelocationPreference(str, Enum):
    YES = "Yes"
    NO = "No"
    REMOTE = "Open to remote"


class HoursPerDay(str, Enum):
    ONE = "1hr"
    TWO_TO_THREE = "2–3hr"
    FOUR_PLUS = "4+hr"


class BudgetInr(str, Enum):
    ZERO = "0"
    LOW = "0–2k"
    MID = "2–10k"
    HIGH = "10k+"


class InterestArea(str, Enum):
    MACHINE_LEARNING = "Machine Learning"
    DATA_SCIENCE = "Data Science"
    UI_UX_DESIGN = "UI/UX Design"
    WEB_DEV = "Web Dev"
    MOBILE_DEV = "Mobile Dev"
    CYBERSECURITY = "Cybersecurity"
    CLOUD_DEVOPS = "Cloud/DevOps"
    PRODUCT_MANAGEMENT = "Product Management"
    GAME_DEV = "Game Dev"
    HARDWARE = "Hardware"
    ROBOTICS = "Robotics"
    BUSINESS_ANALYST = "Business Analyst"
    FINANCE_ACCOUNTING = "Finance & Accounting"
    DIGITAL_MARKETING = "Digital Marketing"
    OTHER = "Other"


class SkillRating(str, Enum):
    HEARD_OF = "heard of it"
    WITH_HELP = "can use with help"
    SOLO = "can build solo"


class PrimaryGoal(str, Enum):
    FIRST_JOB = "First job"
    INTERNSHIP = "Internship"
    HIGHER_STUDIES = "Higher studies"
    STARTUP = "Startup"
    FREELANCING = "Freelancing"


class TargetSalaryBand(str, Enum):
    BAND_3_6 = "₹3–6 LPA"
    BAND_6_12 = "₹6–12 LPA"
    BAND_12_PLUS = "₹12+ LPA"


class TimelineToJob(str, Enum):
    THREE_MONTHS = "3 months"
    SIX_MONTHS = "6 months"
    TWELVE_MONTHS = "12 months"
    TWO_PLUS_YEARS = "2+ years"


LikertScore = Annotated[int, Field(ge=1, le=5)]


class ProfileBasics(BaseModel):
    year_of_study: YearOfStudy
    stream: TechStream
    college_tier: CollegeTier
    cgpa_band: CgpaBand
    city: str = Field(min_length=1, max_length=100)
    relocation: RelocationPreference
    hours_per_day: HoursPerDay
    budget_inr: BudgetInr


class SelfAssessment(BaseModel):
    coding_comfort: LikertScore
    math_logic_comfort: LikertScore
    communication_comfort: LikertScore
    design_visual_comfort: LikertScore
    people_sales_comfort: LikertScore
    detail_patience: LikertScore
    interest_bias: list[InterestArea] = Field(min_length=1, max_length=3)
    commitment_slider: Annotated[int, Field(ge=0, le=10)]

    @field_validator("interest_bias")
    @classmethod
    def unique_interests(cls, values: list[InterestArea]) -> list[InterestArea]:
        if len(values) != len(set(values)):
            raise ValueError("interest_bias entries must be unique")
        return values


class StudentSkill(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    rating: SkillRating


class SkillsAndGoals(BaseModel):
    skills: list[StudentSkill] = Field(default_factory=list)
    projects_text: str | None = Field(default=None, max_length=2000)
    github_url: HttpUrl | None = None
    has_internship: bool
    internship_brief: str | None = Field(default=None, max_length=500)
    primary_goal: PrimaryGoal
    target_salary_band: TargetSalaryBand
    timeline_to_job: TimelineToJob
    dream_companies: list[str] = Field(default_factory=list, max_length=3)

    @field_validator("dream_companies")
    @classmethod
    def trim_companies(cls, values: list[str]) -> list[str]:
        return [company.strip() for company in values if company.strip()]


class IntakeForm(BaseModel):
    profile: ProfileBasics
    self_assessment: SelfAssessment
    skills_and_goals: SkillsAndGoals

File: apps/web/src/lib/schemas/intake.ts
import { z } from "zod";

export const yearOfStudySchema = z.enum([
  "1st",
  "2nd",
  "3rd",
  "4th",
  "Dropout",
  "Gap year",
]);

export const techStreamSchema = z.enum(["CS", "IT", "ECE", "EE"]);

export const collegeTierSchema = z.enum([
  "Tier 1 (IIT/NIT/IIIT)",
  "Tier 2",
  "Tier 3",
]);

export const cgpaBandSchema = z.enum([
  "<6.5",
  "6.5–7.5",
  "7.5–8.5",
  "8.5+",
]);

export const relocationSchema = z.enum(["Yes", "No", "Open to remote"]);
export const hoursPerDaySchema = z.enum(["1hr", "2–3hr", "4+hr"]);
export const budgetInrSchema = z.enum(["0", "0–2k", "2–10k", "10k+"]);

export const interestAreaSchema = z.enum([
  "Machine Learning",
  "Data Science",
  "UI/UX Design",
  "Web Dev",
  "Mobile Dev",
  "Cybersecurity",
  "Cloud/DevOps",
  "Product Management",
  "Game Dev",
  "Hardware",
  "Robotics",
  "Business Analyst",
  "Other",
]);

export const skillRatingSchema = z.enum([
  "heard of it",
  "can use with help",
  "can build solo",
]);

export const primaryGoalSchema = z.enum([
  "First job",
  "Internship",
  "Higher studies",
  "Startup",
  "Freelancing",
]);

export const targetSalaryBandSchema = z.enum([
  "₹3–6 LPA",
  "₹6–12 LPA",
  "₹12+ LPA",
]);

export const timelineToJobSchema = z.enum([
  "3 months",
  "6 months",
  "12 months",
  "2+ years",
]);

export const likertSchema = z.number().int().min(1).max(5);

export const profileBasicsSchema = z.object({
  year_of_study: yearOfStudySchema,
  stream: techStreamSchema,
  college_tier: collegeTierSchema,
  cgpa_band: cgpaBandSchema,
  city: z.string().min(1).max(100),
  relocation: relocationSchema,
  hours_per_day: hoursPerDaySchema,
  budget_inr: budgetInrSchema,
});

export const selfAssessmentSchema = z
  .object({
    coding_comfort: likertSchema,
    math_logic_comfort: likertSchema,
    communication_comfort: likertSchema,
    design_visual_comfort: likertSchema,
    people_sales_comfort: likertSchema,
    detail_patience: likertSchema,
    interest_bias: z.array(interestAreaSchema).min(1).max(3),
    commitment_slider: z.number().int().min(0).max(10),
  })
  .refine(
    (data) => new Set(data.interest_bias).size === data.interest_bias.length,
    { message: "Pick each interest only once", path: ["interest_bias"] },
  );

export const studentSkillSchema = z.object({
  name: z.string().min(1).max(80),
  rating: skillRatingSchema,
});

const optionalUrl = z.preprocess(
  (v) => (v === "" || v === null || v === undefined ? undefined : v),
  z.string().url().optional(),
);

const optionalText = z.preprocess(
  (v) => (v === "" || v === undefined ? null : v),
  z.string().max(2000).nullable().optional(),
);

export const skillsAndGoalsSchema = z.object({
  skills: z.array(studentSkillSchema).default([]),
  projects_text: optionalText,
  github_url: optionalUrl,
  has_internship: z.boolean(),
  internship_brief: z.preprocess(
    (v) => (v === "" || v === undefined ? null : v),
    z.string().max(500).nullable().optional(),
  ),
  primary_goal: primaryGoalSchema,
  target_salary_band: targetSalaryBandSchema,
  timeline_to_job: timelineToJobSchema,
  dream_companies: z.array(z.string()).max(3).default([]),
});

export const intakeFormSchema = z.object({
  profile: profileBasicsSchema,
  self_assessment: selfAssessmentSchema,
  skills_and_goals: skillsAndGoalsSchema,
});

export type YearOfStudy = z.infer<typeof yearOfStudySchema>;
export type TechStream = z.infer<typeof techStreamSchema>;
export type InterestArea = z.infer<typeof interestAreaSchema>;
export type IntakeForm = z.infer<typeof intakeFormSchema>;
export type ProfileBasics = z.infer<typeof profileBasicsSchema>;
export type SelfAssessment = z.infer<typeof selfAssessmentSchema>;
export type SkillsAndGoals = z.infer<typeof skillsAndGoalsSchema>;
export type StudentSkill = z.infer<typeof studentSkillSchema>;
